runtime/eventlog.c

```diff
@@ -0,0 +1,405 @@
+{
+  CAMLassert(!evbuf);
+  evbuf = caml_stat_alloc_noexc(sizeof(*evbuf));
+
+  if (evbuf == NULL)
+    caml_fatal_error("eventlog: could not allocate event buffer");
+
+  evbuf->ev_generated = 0;
+}
+
+static void setup_eventlog_file()
+{
+  CAMLparam0();
+  char_os *filename = NULL;
+  char_os *eventlog_filename;
+  CAMLlocal1 (tmp);
+
+  eventlog_filename = caml_secure_getenv(T("OCAML_EVENTLOG_FILE"));
+  if (eventlog_filename) {
+    char *input =  caml_stat_strdup_of_os(eventlog_filename);
+
+    if (input != NULL) {
+      tmp = caml_alloc_sprintf("%s.%d.eventlog",
+                               input, Caml_state->eventlog_startup_pid);
+      caml_stat_free(input);
+      filename = caml_stat_strdup_to_os(String_val(tmp));
+    }
+
+  } else {
+    tmp = caml_alloc_sprintf("caml-eventlog-%d",
+                             Caml_state->eventlog_startup_pid);
+    filename = caml_stat_strdup_to_os(String_val(tmp));
+  }
+
+  if (filename) {
+    Caml_state->eventlog_out = fopen_os(filename, T("wb"));
+    caml_stat_free(filename);
+  }
+
+  if (Caml_state->eventlog_out) {
+    int ret =  fwrite(&header, sizeof(struct ctf_stream_header),
+                      1, Caml_state->eventlog_out);
+    if (ret != 1)
+      caml_eventlog_disable();
+    fflush(Caml_state->eventlog_out);
+  } else {
+    caml_fatal_error("eventlog: could not open trace for writing");
+  }
+
+  CAMLreturn0;
+}
+
+#define FWRITE_EV(item, size) \
+  if (fwrite(item, size, 1, out) != 1) \
+    goto fwrite_failure;
+
+static void flush_events(FILE* out, struct event_buffer* eb)
+{
+  uintnat i;
+  uint64_t flush_duration;
+  uintnat n = eb->ev_generated;
+
+  struct ctf_event_header ev_flush;
+  ev_flush.id = EV_FLUSH;
+  ev_flush.timestamp = time_counter() -
+                        Caml_state->eventlog_startup_timestamp;
+  ev_flush.pid = Caml_state->eventlog_startup_pid;
+
+  for (i = 0; i < n; i++) {
+    struct event ev = eb->events[i];
+    ev.header.pid = Caml_state->eventlog_startup_pid;
+
+    FWRITE_EV(&ev.header, sizeof(struct ctf_event_header));
+
+    switch (ev.header.id)
+    {
+    case EV_ENTRY:
+      FWRITE_EV(&ev.phase, sizeof(uint16_t));
+      break;
+    case EV_EXIT:
+      FWRITE_EV(&ev.phase, sizeof(uint16_t));
+      break;
+    case EV_COUNTER:
+      FWRITE_EV(&ev.count, sizeof(uint64_t));
+      FWRITE_EV(&ev.counter_kind, sizeof(uint16_t));
+      break;
+    case EV_ALLOC:
+      FWRITE_EV(&ev.count, sizeof(uint64_t));
+      FWRITE_EV(&ev.alloc_bucket, sizeof(uint8_t));
+      break;
+    default:
+      break;
+    }
+  }
+
+  flush_duration =
+    (time_counter() - Caml_state->eventlog_startup_timestamp) -
+    ev_flush.timestamp;
+
+  FWRITE_EV(&ev_flush, sizeof(struct ctf_event_header));
+  FWRITE_EV(&flush_duration, sizeof(uint64_t));
+
+  return;
+
+ fwrite_failure:
+  /* on event flush failure, shut down eventlog. */
+  fprintf(stderr,
+          "eventlog: error while writing trace file, disabling eventlog");
```

Comment:
You should probably use the `caml_runtime_warnings_active` function here (see `io.c` for an example) and at any rate you need a `\n` at the end of the string.